cmd/disk-cache.go

```
@@ -0,0 +1,477 @@
+package cmd
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/minio/minio/pkg/disk"
+	"github.com/minio/minio/pkg/hash"
+)
+
+// Disk caching
+type diskCache struct {
+	dir string // caching directory (disk path in the --cache-policy file)
+	// Max disk usage limit (--cache-max) in percent. If the disk usage crosses this
+	// limit then we stop caching.
+	maxUsage int
+
+	// creation-time of the cache
+	createTime time.Time
+	// cache database
+	db ObjectLayer
+
+	// purge() listens on this channel to start the cache-purge process
+	purgeChan chan struct{}
+}
+
+// cache policy for prefix
+type cachePolicy struct {
+	cachePrefix string
+}
+
+// mapping of disk path to diskCache object
+type diskToCacheMap map[string]*diskCache
+
+// mapping of cache policy to disk path
+type policyToDiskMap map[cachePolicy]string
+
+// Returns if the disk usage is low.
+// Disk usage is low if usage is < 80% of maxUsage
+// Ex. for a 100GB disk, if maxUsage is configured as 70% then maxUsage is 70G
+// hence disk usage is low if the disk usage is less than 56G (because 80% of 70G is 56G)
+func (c diskCache) diskUsageLow() bool {
+	minUsage := c.maxUsage * 80 / 100
+
+	di, err := disk.GetInfo(c.dir)
+	if err != nil {
+		errorIf(err, "Error getting disk information on %s", c.dir)
+		return false
+	}
+	usedPercent := (di.Total - di.Free) * 100 / di.Total
+	return int(usedPercent) < minUsage
+}
+
+// Return if the disk usage is high.
+// Disk usage is high if disk used is > maxUsage
+func (c diskCache) diskUsageHigh() bool {
+	di, err := disk.GetInfo(c.dir)
+	if err != nil {
+		return true
+	}
+	usedPercent := (di.Total - di.Free) * 100 / di.Total
+	return int(usedPercent) > c.maxUsage
+}
+
+// Returns if size space can be allocated without exceeding
+// max disk usage
+func (c diskCache) diskAvailable(size int64) bool {
+	di, err := disk.GetInfo(c.dir)
+	if err != nil {
+		return true
```

eating the error is a bad idea.